Past instance notification

saarasat opened the pull request:
# Description

- If the course instance has ended and enrolling has already closed, the page visitor will be told that the course has already ended. 
- If there is an instance of the same course, that has a later ending time than now, and that instance is visible for that user, they will be given a link to that instance. 
- If there are several instances that fit this description, the latest is given.
- Even if the course has already ended, but the enrolling would for some reason be open, the enroll_form is shown.  
- Front page and enroll-page of the course show the link to the latest instance of the course. 
- Other course-pages show only the text "This course has already ended" to save some database-queries
(If the link is wanted for other pages as well, most likely the best option would be to move the alert to the course_base.html, that seems to work as well and is easy to move around. That however will cause an extra query for old instances. If this issue is considered more important and worth using extra time, a cache entry could be probably implemented.)

Fixes #554

**Front page of the course**:

![image](https://user-images.githubusercontent.com/43036333/87779880-39b25880-c836-11ea-8fb9-dd20ecdc4e15.png)


**Content-pages of the course**

![image](https://user-images.githubusercontent.com/43036333/87779919-4c2c9200-c836-11ea-9fc8-da44ed0dd1ce.png)


# Testing

Unit-tests and selenium-tests pass once this has been merged: https://github.com/apluslms/a-plus/pull/621

The changes have been tested as a logged in user, a student and a staff member with past courses that have no future instances and with courses that have future instances. Both hidden ones and ones visible to students. Also the case where a future instance exist and where the enrolling has not yet started has been tested. 

The changes do not change the way the content is shown to the user, only limit the showing of the enroll_form to courses where the enrolling is still .  
 
o1 has other implementations of this, which would become unnecessary now: 
https://plus.cs.aalto.fi/o1/2018/

# Have you updated the README or other relevant documentation?

Not relevant here.

# Is it [Done](https://wiki.aalto.fi/display/EDIT/Definition+of+Done)?

- [ ] I (developer) have created unit tests and the tests pass
- [ ] I (developer) have created functional tests (Selenium tests) if applicable
- [x] I (developer) have tested the changes manually
- [ ] Reviewer has finished the code review
- [ ] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

raphendyr commented on course/views.py:
```diff
@@ -106,8 +106,23 @@ def get(self, request, *args, **kwargs):
+                    ending_time__gte=now,
+                    course=course
+                ).latest('starting_time')
+            except ObjectDoesNotExist:
```
You can use the specific exception `CourseInstance.DoesNotExist`.

raphendyr commented on course/views.py:
```diff
@@ -106,8 +106,23 @@ def get(self, request, *args, **kwargs):
+        return super().get(request,
+                            *args,
+                            **kwargs,
+                            later_instance=later_instance)
```
If the `later_instance` exists in the `kwargs`, then this line will crash. It might be a good idea to always overwrite the field using line `kwargs['later_instance'] = later_instance`.

raphendyr commented on course/views.py:
```diff
@@ -106,8 +106,23 @@ def get(self, request, *args, **kwargs):
+            try:
+                later_instance = CourseInstance.objects.get_visible(request.user).filter(
+                    ending_time__gte=now,
+                    course=course
```
The line is missing the comma at the end.

raphendyr replied:
In addition, as this chains multiple methods, it might be readable if every methods starts a new line. Hence, maybe:

```python
later_instance = (
    CourseInstance.objects
        .get_visible(request.user)
        .filter(course=course, ending_time__gte=now)
        .latest('starting_time')
)
```

saarasat commented:
> I'm wondering about the appearance.
> 
> First, I wonder if this is prominent enough. The text matches the size and style of nearby text, so maybe there is a risk it gets skipped over.
> 
> My other concern is that if someone follows a link to, for example, heading or an exercise further down the page, they completely miss this even though they might benefit from it. Have you considered trying a sticky version?

That is a valid point. Adding stickiness and pinning it eg. on the top or the bottom of the page might work. I'll try that!

saarasat commented:
I now made the improvements to the python as @raphendyr suggested and changed the alert to a sticky one like @mtthwhggns proposed. I changed the coloring of the alert, as when it was blue, it got mixed with the colors of exercises when scrolling down. Now it should be noticeable but not too over-powering or annoying. There is no default-alert in bs3, but I added styles here for that. In case the switch is done to bs4 at some point, then things should work :)

raphendyr commented on course/templates/course/course_base.html:
```diff
@@ -60,6 +60,21 @@ <h4>{% trans "Calendar" %}</h4>
+                        {% trans "The latest version of the course can be found at: " %}
+                        <a class="alert-link" href="{{ later_instance|url }}">
+                            <span lang="{{ later_instance.language }}">
+                                {{ later_instance.course.name|parse_localization }}: {{ later_instance.instance_name|parse_localization }}
```
Hmmm. I'm not sure if colon is the best separator for combining name and instance name. However, it appears to be the one that is used in those few locations where these are combined. I used format `name (instance)` in MOOC-Jutut, but I'm not sure if it would be better or not.

If `parse_localization` would handle multiple substitution, then we could say something like `{{ later_instance|parse_localization }}`, which would use `CourseInstance.__str__` (which uses format `name: instance`). If we did something like that, then the format could be changed in a single location.

saarasat replied:
I copied this format from the top nav, there the instance is listed as "course: instance". Though the thing that is missing here is the course code, but I thought that with that one it becomes a bit messy looking. The course code is already listed three times on the page as it is. 

![image](https://user-images.githubusercontent.com/43036333/87956804-21ea0700-cab8-11ea-8aba-e0c861dce65e.png)

raphendyr replied:
Hmm.. I think we should visit the format if we touch that menu. In addition, I guess Jutut should use the same format too...

raphendyr commented on course/templates/course/course_base.html:
```diff
@@ -60,6 +60,21 @@ <h4>{% trans "Calendar" %}</h4>
     <div class="col-sm-10" id="course-content">
+        {% if instance.is_past and instance.has_enrollment_closed %}
+            <div class="alert sticky-alert alert-default">
+                {% trans "This course has already ended. " %}
```
The space at the end of the string doesn't seem to be importart. The new line will tell browsers to use a space between text and the next element.

saarasat replied:
Oh nice, good to know. The space has now been removed.

raphendyr commented on course/templates/course/course_base.html:
```diff
@@ -60,6 +60,21 @@ <h4>{% trans "Calendar" %}</h4>
+                {% trans "This course has already ended. " %}
+                {% if later_instance %}
+                    <p>
+                        {% trans "The latest version of the course can be found at: " %}
```
> The space at the end of the string doesn't seem to be importart. The new line will tell browsers to use a space between text and the next element.

saarasat replied:
Same here

raphendyr commented on course/views.py:
```diff
@@ -106,8 +106,24 @@ def get(self, request, *args, **kwargs):
+        later_instance = None
+
+        if self.instance.is_past:
+            course = self.instance.course
```
I think there is also `self.course`. If so, then I think this "alias" might not be so important. Same consideration can be used for the `now`, which is used only once.

saarasat replied:
There seemed to be that alias, so I cleared that now too

raphendyr commented:
I think it's good too. You could rebase and clean the history now, thanks.

saarasat commented:
Rebased, squashed, commit message checked, tests run and the code checked locally one more time.

raphendyr commented on course/templates/course/_enroll_form.html:
```diff
@@ -1,43 +1,43 @@
+					<div>
+						<small>
+							{% blocktrans with end=instance.enrollment_ending_time %}
+							The course enrollment is open until {{ end }}.
```
Did these block trans edits trigger a change in the translation file? I think you could add the `trimmed` option and make sure the translations still work.

saarasat replied:
Great pick. Did not realise that fixing indentation might break old translations where the trimmed-attribute has not been added. 

I now added trimmed to both, compiled the messages and checked that they worked

saarasat commented on course/views.py:
```diff
@@ -106,8 +106,22 @@ def get(self, request, *args, **kwargs):
+            try:
+                later_instance = (
+                    CourseInstance.objects
+                        .get_visible(request.user)
```
Since `request` is an argument of the function, it can be accessed straightaway, without using the `self`. Works both ways though, with the `self` as well.

saarasat commented on course/templates/course/course.html:
```diff
@@ -3,6 +3,7 @@
 {% load exercise %}
 {% load news %}
 {% load apps %}
+{% load course %}
```
Oh, I think that was left there by accident, now that the implementation was different from what I originally tried. I removed that bit, good catch :)

saarasat commented:
> Do you think we should limit the enrolment ending time according to the ending time of the course? @raphendyr Maybe you can explain why this behaviour?

At least its seems to be possible have enrolling open even though the course has ended, with the current way the instances are set. So I thought I should cover that case as well, thus that the enroll_form is shown, if the enrolling is still actually used even though the course has already ended. How much this option is used though, I am not sure.

raphendyr commented:
> Do you think we should limit the enrolment ending time according to the ending time of the course? @raphendyr Maybe you can explain why this behaviour?

Of top of my head, it soulds stranges to be able to enroll after the course has ended. It's possible that there has just been assumption that it's always configured so. @Mankro did revisit those time fields and has a better knowledge how they probably should work.

saarasat commented:
@jaguarfi Just let me know if there is something here then that you'd like me to check or revisit, or whether this can be merged :)

jaguarfi commented:
@saarasat could you create a new issue to verify that the enrollment time behaves according to the duration of the course? So that we can eventually verify whether or not this is the expected behaviour.

saarasat commented:
> @saarasat could you create a new issue to verify that the enrollment time behaves according to the duration of the course? So that we can eventually verify whether or not this is the expected behaviour.

And done https://github.com/apluslms/a-plus/issues/636

saarasat commented:
And just rebased on top of yesterday's latest commit.

jaguarfi commented:
Terrific :100:
